### PR TITLE
scripts: tests: Blackbox test expansion - coverage

### PR DESCRIPTION
**Pull request by KamilxPaszkiet**
Adds tests related to coverage flags:
--gcov-tool
--coverage-platform

### CODE REVIEW

File: scripts/tests/twister_blackbox/test_coverage.py [hunk 1]
```diff
@@ -298,3 +312,83 @@ def test_coverage_tool(self, capfd, caplog, test_path, test_platforms, out_path,
...
+
+    @pytest.mark.parametrize(
+        'test_path, test_platforms, expected_content',
+        TESTDATA_6,
+        ids=[
+            'missing tool'
+        ]
+    )
+    def test_gcov_tool(self, capfd, caplog, test_path, test_platforms, out_path, expected_content):
+        args = ['--outdir', out_path,'-i', '-T', test_path] + \
```
Comment (LukaszMrugala) [inline scripts/tests/twister_blackbox/test_coverage.py:324]: No space after the comma.

File: scripts/tests/twister_blackbox/test_coverage.py [hunk 2]
```diff
@@ -130,6 +130,20 @@ class TestCoverage:
...
+            os.path.join(TEST_DATA, 'tests', 'dummy', 'agnostic', 'group2'),
+            ['qemu_x86'],
+            ['The specified file does not exist.',r'\[Errno 13\] Permission denied:'],
+        )
+    ]
+    TESTDATA_7 = [
+        (
+            os.path.join(TEST_DATA, 'tests', 'dummy', 'agnostic', 'group2'),
+            ['qemu_x86_64','qemu_x86'],
+            ['qemu_x86_64','qemu_x86',['qemu_x86_64','qemu_x86']],
```
Comment (LukaszMrugala) [inline scripts/tests/twister_blackbox/test_coverage.py:144]: No space after the comma.

File: scripts/tests/twister_blackbox/test_coverage.py [hunk 3]
```diff
@@ -130,6 +130,20 @@ class TestCoverage:
             'Running lcov --gcov-tool'
         )
     ]
+    TESTDATA_6 = [
+        (
+            os.path.join(TEST_DATA, 'tests', 'dummy', 'agnostic', 'group2'),
+            ['qemu_x86'],
+            ['The specified file does not exist.',r'\[Errno 13\] Permission denied:'],
```
Comment (LukaszMrugala) [inline scripts/tests/twister_blackbox/test_coverage.py:137]: No space after the comma.

File: scripts/tests/twister_blackbox/test_coverage.py [hunk 4]
```diff
@@ -298,3 +312,83 @@ def test_coverage_tool(self, capfd, caplog, test_path, test_platforms, out_path,
...
+            self.loader.exec_module(self.twister_module)
+
+        out, err = capfd.readouterr()
+        sys.stdout.write(out)
+        sys.stderr.write(err)
+
+        assert str(sys_exit.value) == '1'
+        for line in expected_content:
+            result = re.search(line, err)
+            assert result, 'unknown error'
```
Comment (LukaszMrugala) [inline scripts/tests/twister_blackbox/test_coverage.py:341]: Something more informative, like 'Different error than the one expected' would be better.

File: scripts/tests/twister_blackbox/test_coverage.py [hunk 5]
```diff
@@ -298,3 +312,83 @@ def test_coverage_tool(self, capfd, caplog, test_path, test_platforms, out_path,
...
+                pytest.raises(SystemExit) as sys_exit:
+            self.loader.exec_module(self.twister_module)
+
+        out, err = capfd.readouterr()
+        sys.stdout.write(out)
+        sys.stderr.write(err)
+
+        assert str(sys_exit.value) == '1'
+        for line in expected_content:
+            result = re.search(line, err)
```
Comment (LukaszMrugala) [inline scripts/tests/twister_blackbox/test_coverage.py:340]: You could wrap `line` in `re.escape()` if you do not plan on using any RegEx features. Or even just do `assert line in err`

File: scripts/tests/twister_blackbox/test_coverage.py [hunk 6]
```diff
@@ -298,3 +312,83 @@ def test_coverage_tool(self, capfd, caplog, test_path, test_platforms, out_path,
...
+    @pytest.mark.parametrize(
+        'test_path, test_platforms, cov_platform',
+        TESTDATA_7,
+        ids=[
+            'coverage platform'
+        ]
+    )
+
+    def test_coverage_platform(self, capfd, caplog, test_path, test_platforms, out_path, cov_platform):
+        run = []
```
Comment (LukaszMrugala) [inline scripts/tests/twister_blackbox/test_coverage.py:352]: Better to put it after the inner function declaration if not needed here.

File: scripts/tests/twister_blackbox/test_coverage.py [hunk 7]
```diff
@@ -298,3 +312,83 @@ def test_coverage_tool(self, capfd, caplog, test_path, test_platforms, out_path,
...
+            assert result, 'unknown error'
+
+    @pytest.mark.parametrize(
+        'test_path, test_platforms, cov_platform',
+        TESTDATA_7,
+        ids=[
+            'coverage platform'
+        ]
+    )
+
```
Comment (LukaszMrugala) [inline scripts/tests/twister_blackbox/test_coverage.py:350]: Remove empty line before the method decorator

File: scripts/tests/twister_blackbox/test_coverage.py [hunk 8]
```diff
@@ -298,3 +312,83 @@ def test_coverage_tool(self, capfd, caplog, test_path, test_platforms, out_path,
...
+            else:
+                args += ['--coverage-platform', element]
+
+            with mock.patch.object(sys, 'argv', [sys.argv[0]] + args), \
+                    pytest.raises(SystemExit) as sys_exit:
+                self.loader.exec_module(self.twister_module)
+
+            out, err = capfd.readouterr()
+            sys.stdout.write(out)
+            sys.stderr.write(err)
```
Comment (LukaszMrugala) [inline scripts/tests/twister_blackbox/test_coverage.py:384]: Seems unused - remove.

File: scripts/tests/twister_blackbox/test_coverage.py [hunk 9]
```diff
@@ -298,3 +312,83 @@ def test_coverage_tool(self, capfd, caplog, test_path, test_platforms, out_path,
...
         assert re.search(expected_content, caplog.text), f'{cov_tool} line not found'
+
+    @pytest.mark.parametrize(
+        'test_path, test_platforms, expected_content',
+        TESTDATA_6,
+        ids=[
+            'missing tool'
+        ]
+    )
+    def test_gcov_tool(self, capfd, caplog, test_path, test_platforms, out_path, expected_content):
```
Comment (LukaszMrugala) [inline scripts/tests/twister_blackbox/test_coverage.py:323]: `caplog` seems unused, remove.

File: scripts/tests/twister_blackbox/test_coverage.py [hunk 10]
```diff
@@ -298,3 +312,83 @@ def test_coverage_tool(self, capfd, caplog, test_path, test_platforms, out_path,
...
+
+    @pytest.mark.parametrize(
+        'test_path, test_platforms, cov_platform',
+        TESTDATA_7,
+        ids=[
+            'coverage platform'
+        ]
+    )
+
+    def test_coverage_platform(self, capfd, caplog, test_path, test_platforms, out_path, cov_platform):
```
Comment (LukaszMrugala) [inline scripts/tests/twister_blackbox/test_coverage.py:351]: `caplog` seems unused, remove.

File: scripts/tests/twister_blackbox/test_coverage.py [hunk 11]
```diff
@@ -298,3 +312,83 @@ def test_coverage_tool(self, capfd, caplog, test_path, test_platforms, out_path,
...
+        run = []
+        def search_cov():
+            pattern = r'TOTAL\s+(\d+)'
+            files = os.path.join(out_path,'coverage','coverage.txt')
```
Comment (LukaszMrugala) [inline scripts/tests/twister_blackbox/test_coverage.py:355]: Add spaces after commas

Comment (LukaszMrugala) [inline scripts/tests/twister_blackbox/test_coverage.py:355]: `coverage_file_path` seems like a more descriptive variable name

File: scripts/tests/twister_blackbox/test_coverage.py [hunk 12]
```diff
@@ -298,3 +312,83 @@ def test_coverage_tool(self, capfd, caplog, test_path, test_platforms, out_path,
...
+
+            assert str(sys_exit.value) == '0'
+
+            run += [search_cov()]
+
+            capfd.readouterr()
+            clear_log_in_test()
+
+        assert run[2] > run[0]
+        assert run[2] > run[1]
```
Comment (LukaszMrugala) [inline scripts/tests/twister_blackbox/test_coverage.py:394]: Adding assert error strings would be quite user for the end user here. Something like `'Broader coverage platform selection did not result in broader coverage.'`

File: scripts/tests/twister_blackbox/test_coverage.py [hunk 13]
```diff
@@ -298,3 +312,83 @@ def test_coverage_tool(self, capfd, caplog, test_path, test_platforms, out_path,
...
+                data = file.read()
+            match = re.search(pattern, data)
+            if match:
+                total = int(match.group(1))
+                return total
+            else:
+                print("Error, pattern not found")
+
+        for element in cov_platform:
+            args = ['--outdir', out_path,'-i', '-T', test_path] + \
```
Comment (LukaszMrugala) [inline scripts/tests/twister_blackbox/test_coverage.py:366]: Add spaces after commas

File: scripts/tests/twister_blackbox/test_coverage.py [hunk 14]
```diff
@@ -298,3 +312,83 @@ def test_coverage_tool(self, capfd, caplog, test_path, test_platforms, out_path,
...
+            ['--coverage', '--coverage-formats', 'txt', '-v'] + \
```
Comment (LukaszMrugala) [inline scripts/tests/twister_blackbox/test_coverage.py:367]: Mismatched indentation?

File: scripts/tests/twister_blackbox/test_coverage.py [hunk 15]
```diff
@@ -298,3 +312,83 @@ def test_coverage_tool(self, capfd, caplog, test_path, test_platforms, out_path,
...
+            pattern = r'TOTAL\s+(\d+)'
+            files = os.path.join(out_path,'coverage','coverage.txt')
+            with open(files, 'r') as file:
+                data = file.read()
+            match = re.search(pattern, data)
+            if match:
+                total = int(match.group(1))
+                return total
+            else:
+                print("Error, pattern not found")
```
Comment (LukaszMrugala) [inline scripts/tests/twister_blackbox/test_coverage.py:363]: Irregular quotation marks - the only `"`s in the whole diff. Change to `'`.

**Review by LukaszMrugala — CHANGES_REQUESTED** [review]

File: scripts/tests/twister_blackbox/test_coverage.py (hunk 1 above)
Comment (KamilxPaszkiet) [reply scripts/tests/twister_blackbox/test_coverage.py]: done

File: scripts/tests/twister_blackbox/test_coverage.py (hunk 2 above)
Comment (KamilxPaszkiet) [reply scripts/tests/twister_blackbox/test_coverage.py]: done

File: scripts/tests/twister_blackbox/test_coverage.py (hunk 3 above)
Comment (KamilxPaszkiet) [reply scripts/tests/twister_blackbox/test_coverage.py]: done

File: scripts/tests/twister_blackbox/test_coverage.py (hunk 4 above)
Comment (KamilxPaszkiet) [reply scripts/tests/twister_blackbox/test_coverage.py]: done

File: scripts/tests/twister_blackbox/test_coverage.py (hunk 6 above)
Comment (KamilxPaszkiet) [reply scripts/tests/twister_blackbox/test_coverage.py]: done

File: scripts/tests/twister_blackbox/test_coverage.py (hunk 7 above)
Comment (KamilxPaszkiet) [reply scripts/tests/twister_blackbox/test_coverage.py]: done

File: scripts/tests/twister_blackbox/test_coverage.py (hunk 8 above)
Comment (KamilxPaszkiet) [reply scripts/tests/twister_blackbox/test_coverage.py]: done

File: scripts/tests/twister_blackbox/test_coverage.py (hunk 9 above)
Comment (KamilxPaszkiet) [reply scripts/tests/twister_blackbox/test_coverage.py]: done

File: scripts/tests/twister_blackbox/test_coverage.py (hunk 10 above)
Comment (KamilxPaszkiet) [reply scripts/tests/twister_blackbox/test_coverage.py]: done

File: scripts/tests/twister_blackbox/test_coverage.py (hunk 11 above)
Comment (KamilxPaszkiet) [reply scripts/tests/twister_blackbox/test_coverage.py]: done

Comment (KamilxPaszkiet) [reply scripts/tests/twister_blackbox/test_coverage.py]: done

File: scripts/tests/twister_blackbox/test_coverage.py (hunk 12 above)
Comment (KamilxPaszkiet) [reply scripts/tests/twister_blackbox/test_coverage.py]: done

File: scripts/tests/twister_blackbox/test_coverage.py (hunk 13 above)
Comment (KamilxPaszkiet) [reply scripts/tests/twister_blackbox/test_coverage.py]: done

File: scripts/tests/twister_blackbox/test_coverage.py (hunk 14 above)
Comment (KamilxPaszkiet) [reply scripts/tests/twister_blackbox/test_coverage.py]: done

File: scripts/tests/twister_blackbox/test_coverage.py (hunk 15 above)
Comment (KamilxPaszkiet) [reply scripts/tests/twister_blackbox/test_coverage.py]: done

**Review by nashif — APPROVED** [review]

**Review by mdubielx — APPROVED** [review]

**Review by mdubielx — APPROVED** [review]

**Review by LukaszMrugala — APPROVED** [review]

**Pull request merged**